### PR TITLE
Run SMT solver on requested queries

### PR DESCRIPTION
**Pull request by leonardoalt**
and rerun compiler with given responses

### CODE REVIEW

**Comment by axic** [comment]
There are 3 PRs fixing this and this doesn't work on Node 4 :)

**Comment by leonardoalt** [comment]
What's the workaround then?

**Comment by axic** [comment]
#105 is the solution.

#127 and #246 do the same change as you.

**Comment by axic** [comment]
It seems however that your change is somewhat a 3rd option which indeed works.

**Comment by axic** [comment]
Oh works because ti doesn't have the test! :)

**Comment by axic** [comment]
Can you try adding your version to #127 (just push a commit) ?

**Comment by leonardoalt** [comment]
It should be the same, shouldn't it?

**Comment by leonardoalt** [comment]
Pushed there

File: solcjs [hunk 1]
```diff
@@ -6,6 +6,7 @@ var originalUncaughtExceptionListeners = process.listeners("uncaughtException");
 var fs = require('fs-extra');
 var path = require('path');
 var solc = require('./index.js');
+var SMTChecker = require('./SMTChecker.js');
```
Comment (axic) [inline solcjs]: This repo uses lowercase filenames.

File: SMTChecker.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+var fs = require('fs-extra');
+var tmp = require('tmp');
+
+// This function checks the standard JSON output for auxiliaryInputRequested,
+// where smtlib2queries represent the queries created by the SMTChecker.
+// The function runs an SMT solver on each query and compiles the input
+// again providing the answers given by the solvers.
+function handleSMTQueries(input, output) {
+  var outputJSON = JSON.parse(output);
+  var auxInputReq = outputJSON["auxiliaryInputRequested"];
```
Comment (axic) [inline SMTChecker.js]: Can just use `outputJSON.auxiliarInputRequested` here and below.

File: SMTChecker.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  // reported if solvers disagree (i.e. SAT vs UNSAT).
+  var potentialSolvers = ["z3", "cvc4"];
+  var solvers = [];
+  for (var solver in potentialSolvers)
+    if (commandExistsSync(potentialSolvers[solver])) {
+      solvers.push(potentialSolvers[solver]);
+      break;
+    }
+  if (solvers.length == 0) {
+    console.log("No SMT solver available. Assertion checking will not be performed.");
```
[the remark this answers is not in the document]
Comment (leonardoalt) [reply SMTChecker.js]: Which function should I use?

**Comment by axic** [comment]
This is my proposal:
```
// Returns `null` if nothing is to be solved or a new `input` if it needs to be run again.
function handleSMTQueries(input, output, solver) {
  if (!output.auxiliaryInputRequested)
    return null

  // Could also just throw an exception here?
  if (!solverCallback)
    return null

  let queries = output.auxiliaryInputRequested.smtlib2queries
  var responses = {}
  for (var queryKey in queries) {
    responses[queryKey] = solver(queries[queryKey])
  }

  input.auxiliaryInput.smtlib2esponses = responses
  return input
}
```

File: SMTChecker.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  var queries = outputJSON["auxiliaryInputRequested"]["smtlib2queries"];
+  var responses = {};
+  if (!queries || Object.keys(queries) == 0)
+	return output;
+  else {
+    for (var query in queries) {
+      var tmpFile = tmp.fileSync();
+      fs.writeFileSync(tmpFile.name, queries[query]);
+      var solverOutput = execSync(solvers[0] + ' ' + tmpFile.name);
```
[the remark this answers is not in the document]
Comment (leonardoalt) [reply SMTChecker.js]: Why callback? Is it ok to have everything sync?

**Comment by leonardoalt** [comment]
`input.auxiliaryInput.smtlib2esponses = responses`
doesn't work because `input.auxiliaryInput` is still `undefined` at that point

**Comment by axic** [comment]
It is the high level design :)

**Comment by leonardoalt** [comment]
Updated extracting the `solve` part into a sync function

**Comment by leonardoalt** [comment]
not sure about the changes to `solcjs`, it's gotten quite long with the error handling

File: smtchecker.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+// The function runs an SMT solver on each query and adjusts the input for
+// another run.
+// Returns null if no solving is requested.
+function handleSMTQueries (inputJSON, outputJSON) {
+  var auxInputReq = outputJSON.auxiliaryInputRequested;
+  if (!auxInputReq)
+    return null;
+
+  if (solvers.length == 0)
+    throw "No SMT solver available. Assertion checking will not be performed."
```
Comment (axic) [inline smtchecker.js]: What syntax is this?

File: solcjs [hunk 6]
```diff
@@ -52,7 +53,27 @@ function abort (msg) {
 
 if (argv['standard-json']) {
   var input = fs.readFileSync(process.stdin.fd);
-  console.log(solc.compile(input.toString('utf8')))
+  var output = solc.compileStandardWrapper(input.toString('utf8'));
+
+  try {
+    var inputJSON = smtchecker.handleSMTQueries(JSON.parse(input), JSON.parse(output));
+    if (inputJSON)
+      output = solc.compileStandardWrapper(JSON.stringify(inputJSON));
```
Comment (axic) [inline solcjs]: This should not be wrapped in this try-catch.

File: solcjs [hunk 7]
```diff
@@ -52,7 +53,27 @@ function abort (msg) {
 
 if (argv['standard-json']) {
   var input = fs.readFileSync(process.stdin.fd);
-  console.log(solc.compile(input.toString('utf8')))
+  var output = solc.compileStandardWrapper(input.toString('utf8'));
+
+  try {
+    var inputJSON = smtchecker.handleSMTQueries(JSON.parse(input), JSON.parse(output));
```
Comment (axic) [inline solcjs]: Could just use `input` here (to override the one above)

File: package.json [hunk 8]
```diff
@@ -39,11 +39,13 @@
   },
   "homepage": "https://github.com/ethereum/solc-js#readme",
   "dependencies": {
+    "command-exists": "^1.2.8",
     "fs-extra": "^0.30.0",
     "keccak": "^1.0.2",
     "memorystream": "^0.3.1",
     "require-from-string": "^2.0.0",
     "semver": "^5.5.0",
+    "tmp": "0.0.33",
```
Comment (axic) [inline package.json]: Based on this version number this seems to be a non-maintained dependency. We have to be careful what kind of dependencies we take.

File: smtchecker.js (hunk 5 above)
Comment (axic) [reply smtchecker.js]: `throw new Exception("...")`

But better to use `assert("No SMT..")`

File: package.json (hunk 8 above)
Comment (leonardoalt) [reply package.json]: Do you know any other lib for this? When I searched it seemed the most common

File: smtchecker.js (hunk 5 above)
Comment (leonardoalt) [reply smtchecker.js]: but we don't want it to crash, the queries are simply not solved then

Comment (axic) [reply smtchecker.js]: Assert is just a basic exception.

File: package.json (hunk 8 above)
Comment (axic) [reply package.json]: Ok it has a fair amount of users.

**Comment by axic** [comment]
Reordered some things and fixed lint.

@leonardoalt can you double check it works?

Can you create an issue for adding tests? That will need a manual installation of cvc4/z3.

**Comment by leonardoalt** [comment]
Yes, it works!

**Pull request merged**